### PR TITLE
fix: allow optional str int etcc in dataclass

### PR DESCRIPTION
**Pull request by samsja**
# Context 

Original issue : https://github.com/jina-ai/docarray/issues/591

Optional field and docarray dataclass does not play well

Indeed for instance one would except that if I create a Document out of a dataclass with an `Optional[Text]` field with default value `None` that the Document will contain an empty Document in bar (added to the chunk). Atm These chunk Document is not created as shown in this code

```python
from typing import Optional

from docarray import dataclass, Document
from docarray.typing import Text


@dataclass
class WebDoc:
    foo: Text
    bar: Optional[Text] = None


d = Document(WebDoc(foo='hello'))
d.bar.text = 'world' # wont work

assert len(d.chunks) == 2  # fail
assert isinstance(d.bar, Document)  # fail
>>>Traceback (most recent call last):
  File "/home/sami/trash/test.py", line 16, in <module>
    d = Document(WebDoc(foo='hello'))
  File "/home/sami/Documents/workspace/Jina/docarray/docarray/document/__init__.py", line 147, in __init__
    super().__init__(*args, **kwargs)
  File "/home/sami/Documents/workspace/Jina/docarray/docarray/base.py", line 38, in __init__
    self._data = type(self)._from_dataclass(_obj)._data
  File "/home/sami/Documents/workspace/Jina/docarray/docarray/document/mixins/multimodal.py", line 116, in _from_dataclass
    raise TypeError(
TypeError: Unsupported type annotation on field `None`
```

# What this pr do

it allows to do
```python


  @dataclass
  class MultiModalDoc:
      foo: Optional[Text] = None

  d = Document(MultiModalDoc())

  d.foo.text = 'world'  
```
This mean it will create a chunk empty Document. Before it was doing nothing

This work as well with non docarray type.
```python
@dataclass
class MultiModalDoc:
    foo: Optional[str] = None

d = Document(MultiModalDoc())
d.tags['foo'] = 'blabla'
```

### CODE REVIEW

**Comment by JoanFM** [comment]
The explanation of this PR does not make sense, what is bar and what is foo? can u clarify what is the wrong behavior and what is expected after this?

Also this seems a bug and therefore should be a fix:

File: tests/unit/array/test_dataclass.py [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+def test_dataclass_opregetional_value():
+    @dataclass
+    class MultiModalDoc:
+        foo: Text
+
+    d = Document(MultiModalDoc(foo='sfdssf'))
+
+
+def test_dataclass_optional_value():
```
Comment (JoanFM) [inline tests/unit/array/test_dataclass.py:18]: why `bar`? where does this bar come from?

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: tests/unit/array/test_dataclass.py (hunk 1 above)
Comment (samsja) [reply tests/unit/array/test_dataclass.py:18]: my bad it should be foo. This test is not passing atm anyway

Comment (JoanFM) [reply tests/unit/array/test_dataclass.py:10]: adjust PR description

File: docarray/document/mixins/multimodal.py [hunk 2]
```diff
@@ -41,12 +46,48 @@ def _from_dataclass(cls, obj) -> 'Document':
...
+
+            elif field.type in [bytes, Optional[bytes]] and not _is_field(field):
+                tags[key] = (
+                    base64.b64encode(attribute).decode() if attribute else attribute
+                )
+                multi_modal_schema[key] = {
+                    'attribute_type': AttributeType.PRIMITIVE,
+                    'type': field.type.__name__
+                    if not (_is_optional(field.type))
+                    else f'Optional[{_get_optional_subtype(field.type).__name__}]',
```
Comment (samsja) [inline docarray/document/mixins/multimodal.py:69]: Here I am open to suggestion should I just remove the Optional in the type name ?

**Comment by samsja** [comment]
Question, should we allow the same behavior with List ?
```python
@dataclass
class WebDoc:
    foo: List[Optional[Text]] = [None] 
```
or 
```python
@dataclass
class WebDoc:
    foo: Optional[List[Text]] = None
```

**Comment by JoanFM** [comment]
Are we 100% this is the expected behavior? I am not sure if I would expect it not to create the `Optional` document but do it lazily perhaps?

For instance:

```python
from typing import Optional

from docarray import dataclass, Document
from docarray.typing import Text


@dataclass
class WebDoc:
    foo: Text
    bar: Optional[Text] = None


d = Document(WebDoc(foo='hello'))
d.bar.text = 'world' # wont work

assert len(d.chunks) == 2  # fail
assert isinstance(d.bar, Document)  # fail
```

Are we sure that you would expect `d.chunks` to be 2 even before creating the `d.bar`? 

Wouldn't this be more expected?

```python
from typing import Optional

from docarray import dataclass, Document
from docarray.typing import Text


@dataclass
class WebDoc:
    foo: Text
    bar: Optional[Text] = None



d = Document(WebDoc(foo='hello'))

len(d.chunks) to be 1
d.bar not existing?

d.bar.text = 'world' # wont work

assert len(d.chunks) == 2  #SUCCEED
assert isinstance(d.bar, Document)   #SUCCEED

```

I am however not sure this is even possible

**Comment by samsja** [comment]
@JoanFM I am not sure neither.

My though is that when you add a field with the default value, the default value will be passed to the Document at init.

```python
@dataclass
class WebDoc:
    foo: Text = 'foo'

d = Document(WebDoc())
```

under the hood this will just call `Document(text='foo')`

Here I am following the exact same idea and under the hood if the default value is None it will call `Document(text=None)`

**Comment by JoanFM** [comment]
> My though is that when you add a field with the default value, the default value will be passed to the Document at init.

Okey, make sense. Make sure to document clearly

**Pull request closed without merging**